File: py/estimation.py
import numpy as np
import pymc3 as pm
from py.kjma import repair_fraction


def mcmc_sample(obs, time_points, m, theta, n_samples=2000, n_chains=5, v_type='argmin_n'):
    g = 1.  # fix growth rate of radius to 1 unit per time step
    with pm.Model():
        n = pm.Bound(pm.Exponential, lower=0.0, upper=1.0)('n', lam=1., testval=.01)
        sig = pm.Bound(pm.Normal, lower=0)('sig', mu=np.pi, sigma=1, testval=np.pi)
        beta = (sig * n * g**(m-1))**(1./m)
        p = pm.Deterministic('p', repair_fraction(time_points, m, beta, theta))
        obs = pm.Bernoulli('obs', p, observed=obs)

        step = pm.Metropolis()
        cellmd_trace = pm.sample(n_samples, step=step, chains=n_chains, return_inferencedata=False)

    n_p, g_p, sig_p = None, 1., None
    if 'argmin' in v_type.lower():
        if v_type.lower() == 'argmin_n':
            amin = np.argmin(cellmd_trace['n'])
        elif v_type.lower() == 'argmin_sig':
            amin = np.argmin(cellmd_trace['sig'])
        else:
            raise ValueError('Unsuppoted v_type')
        n_p = cellmd_trace['n'][amin]
        sig_p = cellmd_trace['sig'][amin]
    elif v_type.lower() == 'mean':
        n_p = np.mean(cellmd_trace['n'])
        sig_p = np.mean(cellmd_trace['sig'])
    else:
        raise ValueError('Unsupportede v_type')

    return n_p, g_p, sig_p

File: jl/estimation.jl
using Turing, MCMCChains
using Distributions
using Random


@model function nucleation_prob(
    y::Vector{Float64},
    time_points::Vector{Float64}, 
    m::Float64, 
    theta::Float64, 
    exp_decay::Float64=10., 
    norm_sig::Float64=5.
)
    # priors for nucleation rate n and growth speed g
    n_p ~ truncated(Exponential(exp_decay), 0., 1.)
    shape_p ~ truncated(Normal(pi, norm_sig), 0., Inf)
    
    # computaion of beta and probbility p
    beta = (shape_p .* n_p).^(1 ./ m)
    p = repair_fraction(time_points, m, beta, theta)

    # The number of observations and probabilities
    for i in 1:length(y)
        y[i] ~ Bernoulli(p[i])
    end
end

function mcmc_sample(
    observation::Vector{Float64},
    t_time::Vector{Float64}, 
    m::Float64, 
    theta::Float64;
    exp_decay::Float64=1., 
    norm_sig::Float64=1., 
    iterations::Int=1000,
    show_progress::Bool=false
)
    # Create Gibbs sampler for each variable
    alg = Gibbs(MH(:n_p), MH(:shape_p))
    # Start sampling.
    chain = sample(
        nucleation_prob(
            observation, 
            t_time,
            m,
            theta,
            exp_decay,
            norm_sig
        ),
        alg,
        iterations,
        progress=show_progress,
    )
    return chain
end

function fetch_data(chain::MCMCChains.Chains, v_type::String="argmin_n")::Tuple{Float64, Float64, Float64}
    n_p, g_p, sig_p = nothing, nothing, nothing
    if lowercase(v_type) == "mean"
        n_p = mean(chain[:n_p])
        g_p = 1.
        sig_p = mean(chain[:shape_p])
    elseif lowercase(v_type) == "argmin_n"
        amin = argmin(chain[:n_p])
        n_p = chain[:n_p][amin]
        g_p = 1.
        sig_p = chain[:shape_p][amin]
    elseif lowercase(v_type) == "argmax_g"
        amax = argmax(chain[:g_p])
        n_p = chain[:n_p][amax]
        g_p = 1.
        sig_p = chain[:shape_p][amax]
    elseif lowercase(v_type) == "argmin_sig"
        amin = argmin(chain[:shape_p])
        n_p = chain[:n_p][amin]
        g_p = 1.
        sig_p = chain[:shape_p][amin]
    else
        throw(ArgumentError("Passed v_type %s is not accepted. 
        Try mean|argmin_n|argmin_g|argmin_sig" % v_type))
    end
    return n_p, g_p, sig_p
end
        